### ISSUE TITLE
Spring Session will not build in IntelliJ IDEA

### ISSUE THREAD

**Issue by jxblum** [spring-projects/spring-session#1163]
Currently, the _Spring Session_ project does **not** build (compile) cleanly with **IntelliJ IDEA**.  This is primarily caused by _Gradle_ projects organizing _Integration Tests_ into a separate directory location ([for example](https://github.com/spring-projects/spring-session/tree/master/spring-session-data-redis/src/integration-test)).

As a result, for convenience, these _Gradle_ projects typically [depend on](https://github.com/spring-projects/spring-session/blob/master/build.gradle#L3) the [_Spring Build Conventions_]([url](https://github.com/spring-gradle-plugins/spring-build-conventions)) _Gradle_ plugin, which contains [custom logic](https://github.com/spring-gradle-plugins/spring-build-conventions/blob/master/src/main/groovy/io/spring/gradle/convention/IntegrationTestPlugin.groovy) to handle the `integrationTest` task and configuration.

As part of the `integrationTest` definition, both `compile` and `runtime` classpath configurations are [defined](https://github.com/spring-gradle-plugins/spring-build-conventions/blob/master/src/main/groovy/io/spring/gradle/convention/IntegrationTestPlugin.groovy#L57-L62).

When dependencies are defined for one of these `integrationTest` classpath configurations ([for example](https://github.com/spring-projects/spring-session/blob/master/samples/javaconfig/redis/spring-session-sample-javaconfig-redis.gradle#L18)), IDEA fails to successfully build/compile the project because the IDEA module classpath does not contain the same `integrationTest` classpath configuration as is currently [handled for Eclipse](https://github.com/spring-gradle-plugins/spring-build-conventions/blob/master/src/main/groovy/io/spring/gradle/convention/IntegrationTestPlugin.groovy#L119-L123) in the _Spring  Build Conventions_ _Gradle_ plugin.

Therefore, to workaround this problem, a developer must typically duplicate the _Gradle_ module classpath configuration by declaring the `testCompile` property to contain the same dependencies as `integrationTestCompile` ([for example](https://github.com/spring-projects/spring-session-data-geode/blob/2.0.4.RELEASE/samples/javaconfig/gemfire-clientserver/spring-session-sample-javaconfig-gemfire-clientserver.gradle#L21-L23)), even if the `testCompile` would not be necessary otherwise.

This is less than ideal from a project maintenance standpoint, and could potentially (though very unlikely) cause conflicts.

No matter; There is a better solution!

I have [enhanced](https://github.com/spring-gradle-plugins/spring-build-conventions/pull/33) the _Spring Build Conventions_ _Gradle_ Plugin to correctly configure IDEA with the classpath configuration handling `integrationTest`.

**Comment by vpavic** [comment]
@jxblum Which version of IntelliJ are you using? I assume it's 2018.1 as it looks like a problem I reported a while ago in spring-gradle-plugins/spring-build-conventions#28.

This actually turned out to be a problem with IntelliJ tooling support, as the update to 2018.2 resolved the problem for me, and the project builds cleanly now. Can you confirm?

**Comment by jxblum** [comment]
Hi @vpavic - I was running `2017.1.5`.

_Spring Build Conventions_ [PR #33](https://github.com/spring-gradle-plugins/spring-build-conventions/pull/33) that I just submitted today resolves the issue.

Anyway, I just downloaded and installed IJ `2018.2.1`.  Then, I setup the _Spring Session_ project, built it, and I get the same error.  I don't think this particular issue is a problem with IJ tooling.

So, I applied my changes from [PR #33](https://github.com/spring-gradle-plugins/spring-build-conventions/pull/33) once again, and then I am able to build _Spring Session_, even in IJ `2018.2.1`.

Are you certain you are building everything? Including the `samples` along with the `integrationTestSrc`?  (Of course, given Spring Session includes and applies the Gradle plugins from _Spring Build Conventions_ everywhere, so I guess you should be unless you defined explicit compilation excludes in IJ itself).

Thus, you should be getting a compilation error on the `Selenium` dependencies since they are only defined in the `integrationTestCompile` configuration, which IJ does not recognize without the changes I submitted in [PR #33](https://github.com/spring-gradle-plugins/spring-build-conventions/pull/33).

IJ does recognize `integrationTest` source to compile, compliments of the _Spring Build Conventions_ Gadle Plugin, IntegrationTestPlugin.groovy, [here](https://github.com/spring-gradle-plugins/spring-build-conventions/blob/master/src/main/groovy/io/spring/gradle/convention/IntegrationTestPlugin.groovy#L91).

But when you go to compile the sample integrationTest source, you get...

<img width="949" alt="screen shot 2018-08-19 at 11 14 44 pm" src="https://user-images.githubusercontent.com/5247119/44323227-d29b8780-a405-11e8-9450-ba8d2f1ffd27.png">

**Comment by vpavic** [comment]
I don't do any manual intervention or setup at all, the entire project compiles cleanly for me with IntelliJ 2018.2.

After taking a closer look, this one seems different from the issue I linked in my previous comment as that one involved `spring-session-docs` module.

How are you importing the project in IntelliJ? Are you perhaps using Gradle's `idea` task? If yes, that might be the cause as I'm simply importing the project using IntelliJ's support for Gradle, rather than other way around (i.e. using Gradle task to generate IntelliJ project files).

**Comment by jxblum** [comment]
I am import the the "_Project from Existing Sources_" and then select Gradle as the project model.

**Comment by jxblum** [comment]
Also, do you use IJ to build or do you build in IJ with the _Spring Session_ `build.gradle` script (this doesn't count)?  I build everything using IJ's compiler (`Build -> Rebuild Project`).

**Comment by vpavic** [comment]
> I am importing the "_Project from Existing Sources_" and then selecting Gradle as the project model.

> I build everything using IJ's compiler (`Build -> Rebuild Project`).

I do the same for both. Will try to import a fresh clone at some point today to double check and get back to you.

**Comment by jxblum** [comment]
Thanks @vpavic.  I will do the same on my wife's computer tonight as well.  She does not have Java, IJ, Git or the Spring Session project, so I will start everything from scratch. :)

**Comment by rwinch** [comment]
FWIW I was able to import a fresh clone of Spring Session master yesterday using IntelliJ 2018.2.1

**Comment by jxblum** [comment]
What really is the problem of accepting/merging this _Spring Build Conventions_ [PR (#33)](https://github.com/spring-gradle-plugins/spring-build-conventions/pull/33)?  

As I stated arleady in that PR, it has the same effect as [this](https://github.com/spring-gradle-plugins/spring-build-conventions/blob/master/src/main/groovy/io/spring/gradle/convention/IntegrationTestPlugin.groovy#L121), which is required to get Eclipse to work, and I took the required snippet of code from [here](https://docs.gradle.org/current/dsl/org.gradle.plugins.ide.idea.model.IdeaModule.html).

Either way, _Spring Session_ will not build for IJ `2017` and earlier users (possibly `2018.1` and earlier too since @vpavic pointed out that apparently this was not resolved until `2018.2`, which I have yet to experience).  IMO, we should not expect users to upgrade their tooling when the PR #33 that I submitted for _Spring Build Conventions_ resolves this for IJ `2017` and (for me) `2018.2.1`.

Finally, let me ask this question... 

_How does/would IJ knowing anything about the `integrationTestCompile` classpath?

In the example I referenced above, Gradle was showing how to add a "custom" classpath (`performanceTestCompile`) to IJ's TEST scope.  `integrationTestCompile` is only known by Gradle when applying plugins from _Spring Build Conventions_.  How does/would IJ know this?  IJ did not know anything about the `integrationTest` **source** without [help](https://github.com/spring-gradle-plugins/spring-build-conventions/blob/master/src/main/groovy/io/spring/gradle/convention/IntegrationTestPlugin.groovy#L91).

We are not aligned on something.

**Comment by jxblum** [comment]
My steps...

* I am setting this up on my wife's computer (Mac OS X `10.10.5`).  
* I install JDK `1.8.0_181-b13` (the latest from Oracle for Mac). 
* I downloaded and install IJ `2018.2.1` from JetBrains.
* I git cloned _Spring Session_, and from IJ, started with "_Import Project_".
* Next, I choose "_Import project from external model_" and select "_Gradle_".
* I **select** "_Use auto-import_", **unselect** "_Create separate module per source set_" and I **keep** the "_Use default gradle wrapper (recommended)_".
* "_Finish_"

IJ begins downloading `gradle-4.9-bin.zip` and the proceeds to refresh/download the project dependencies.  (IJ continues to download the Internet).

Many minutes later... I am ready build. I do and... FAIL!

<img width="1385" alt="spring-session-idea-compile-failure" src="https://user-images.githubusercontent.com/5247119/44420038-b39a1400-a531-11e8-9312-5bebb58d2924.png">

**Comment by rwinch** [comment]
The reason I am hesitant to make any changes to the build conventions is because we still don't fully understand what the problem is. Why does it work for Vedran and myself but not you? It is critical to answer that question before we make any code changes.

**Comment by jxblum** [comment]
It does not work on my wife's computer either (which had fresh installs of everything: Java, Git, IJ, Spring Session, etc, literally from scratch). I suspect it will not work for other users if you follow my steps in this [comment](https://github.com/spring-projects/spring-session/issues/1163#issuecomment-414767122). And, either way, it most certainly won't work for IJ `2017` (or earlier) users (assuming JetBrains did something special to recognize "custom" `integrationTest` configurations, which I don't believe they did).  This simply not a problem because my computer/IDE is messed up in this case and somehow, everyone else's works.

Again, if you can explain why [this](https://github.com/spring-gradle-plugins/spring-build-conventions/blob/master/src/main/groovy/io/spring/gradle/convention/IntegrationTestPlugin.groovy#L121) is needed for Eclipse, and why this is needed to even recognize the "custom" `integrationTest` **source** directory in IJ (which IJ will not recognize as additional source without this), then you have an answer.

There are many different reasons why it might work.

* Excluding the compile of `integrationTest` **sources** importing `Selenium` classes
* Adding a Global Library for `Selenium` libs in IJ (where the configuration of global libs could have been copied over from an earlier IJ installation, which IJ prompts you to reuse on a fresh/new install of IJ... "use existing configurations from earlier IJ version").
* Building using Gradle inside IJ
* etc...

I am going to try 1 more thing.

Please follow the steps outlined [here](https://github.com/spring-projects/spring-session/issues/1163#issuecomment-414767122).  This will reproduce the issue.

**Comment by vpavic** [comment]
The reason it doesn't work for @jxblum is this:

> **unselect** "_Create separate module per source set_"

Any particular reason why you are disabling module per source set arrangement @jxblum? I've also tried with fresh clone and import and it worked for me, but after unselecting this option myself and refreshing the project, it fails to compile with the same error.

**Comment by jxblum** [comment]
I don't like having a module per source set.  I run into other issues when building other projects with this option selected.  Earlier versions of IJ did not do this and I prefer the older arrangement.

**Comment by jxblum** [comment]
I just verified, and I too can build _Spring Session_ when I use "_Create separate module per source set_" (though this is not my preference).

I also verified that the _Spring Build Conventions_ modifications I made in [PR #33](https://github.com/spring-gradle-plugins/spring-build-conventions/pull/33) will *NOT* affect IJ even  when using "_Create separate module per source set_".

Therefore the changes in [PR #33](https://github.com/spring-gradle-plugins/spring-build-conventions/pull/33) are highly recommended for users not creating separate modules per source set when using IDEA (this was not even an option prior to IJ `2017` I think.  Well, particularly not in `2015' anyway; I missed the `2016` update).
